### PR TITLE
Improve schema building, bug fixes

### PR DESCRIPTION
**Pull request by altro3**
Main changes:
* Added correct processing for optional path variables (routing like this: `"/test{/pathVar1,pathVar2}"`)
* Fixes process oneOf block on field level 
* Added unwrapping allOf block when it possible
* Now the in field for parameters without annotation is determined automatically (`path` in all cases, `query` for get requests)
* Fixed duplicate elements in oneOf and anyOf blocks

Code changes
* Now all tests check final openApi object for more correct checks (found some incorrect tests after it)
* Now all modifications for the OpenAPI object after the analysis of the class structure are moved to the method `postProcessOpenApi`
* Some minor code fixes

### CODE REVIEW

**Review by dstepanov — CHANGES_REQUESTED** [review]
It's hard to understand what is fixed/changed. Can you keep the old tests as they are if modification is not needed and add new ones? Or please comment on every test change and describe why it changed. Thanks!

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiPojoControllerSpec.groovy [hunk 1]
```diff
@@ -366,26 +366,26 @@ class MyBean {}
         petSchema.properties['name'].type == 'string'
         petSchema.properties['name'].description == 'The Pet Name'
 
-        ((MapSchema) petSchema.properties['freeForm']).type == "object"
-        ((MapSchema) petSchema.properties['freeForm']).description == "A free-form object"
-        ((MapSchema) petSchema.properties['freeForm']).getAdditionalProperties() == true
+        petSchema.properties['freeForm'].type == "object"
+        petSchema.properties['freeForm'].description == "A free-form object"
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiPojoControllerSpec.groovy:370]: Now we have just instance of Schema class

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiPathParamRegexSpec.groovy [hunk 2]
```diff
@@ -21,7 +21,7 @@ import io.swagger.v3.oas.annotations.Operation;
 class OpenApiController {
 
     @Operation(summary = "Update tag", description = "Updates an existing tag", tags = "users_tag")
-    @Post("/tags/{tagId: \\\\d+}/{path:.*}{.ext}/update{/id:[a-zA-Z]+}/{+path}{?max,offset}")
+    @Post("/tags/{tagId: \\\\d+}/{path:.*}{.ext}/update/{+path}{?max,offset}{/id:[a-zA-Z]+}")
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiPathParamRegexSpec.groovy:24]: optional path variable must be last

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiParameterSchemaSpec.groovy [hunk 3]
```diff
@@ -80,20 +81,16 @@ package test;
...
 
-import java.time.Period;
-import java.util.Optional;
-
 @Controller("/path")
 class OpenApiController {
 
     @Patch
-    public HttpResponse<String> processSync4(@Parameter(schema = @Schema(type = "string", format = "uuid")) String param) {
+    public HttpResponse<String> processSync4(@Parameter(schema = @Schema(type = "string", format = "uuid")) @QueryValue String param) {
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiParameterSchemaSpec.groovy:93]: Added lost annotation `@QueryValue`

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiParameterSchemaSpec.groovy [hunk 4]
```diff
@@ -28,19 +28,19 @@ class OpenApiController {
...
     public HttpResponse<String> processSync2(
-            @Parameter(ref = "#/components/parameters/MyParam") Optional<Period> period) {
+            @Parameter(schema = @Schema(minLength = 10, maxLength = 20)) @QueryValue Optional<Period> period) {
         return HttpResponse.ok();
     }
 
     @Put
     public HttpResponse<String> processSync3(
-            @Parameter(schema = @Schema(ref = "#/components/schemas/MyParamSchema")) Optional<Period> period) {
+            @Parameter(schema = @Schema(ref = "#/components/schemas/MyParamSchema")) @QueryValue Optional<Period> period) {
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiParameterSchemaSpec.groovy:43]: Added lost `@QueryValue` annotation

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiParameterMappingSpec.groovy [hunk 5]
```diff
@@ -423,25 +423,28 @@ import io.swagger.v3.oas.annotations.enums.*;
...
     @Get("/test6{?bar}")
-    public String test6(@Nullable String bar, String name);
+    String test6(@Nullable String bar, String name);
 
     @Post("/test7")
-    public String test7(String someId, @Nullable String someNotRequired, java.util.Optional<String> someNotRequired2, HttpRequest req, Principal principal, @Body Greeting myBody);
+    String test7(String someId, @Nullable String someNotRequired, java.util.Optional<String> someNotRequired2, HttpRequest req, Principal principal, @Body Greeting myBody);
+
+    @Get("/test8{/pathVar1,pathVar2}")
+    String test8(String pathVar1, String pathVar2, String queryVar, @Nullable String name);
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiParameterMappingSpec.groovy:447]: new test case for optional path vars

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiInheritedPojoControllerSpec.groovy [hunk 6]
```diff
@@ -1594,13 +1594,10 @@ class MyBean {}
         !requestTypeSchema.allOf
 
         exportRequestTypeSchema
-        exportRequestTypeSchema.type == 'object'
-        exportRequestTypeSchema.allOf
-        exportRequestTypeSchema.allOf.size() == 1
-        exportRequestTypeSchema.allOf[0].$ref == '#/components/schemas/RequestType2_4_0'
+        !exportRequestTypeSchema.allOf
+        exportRequestTypeSchema.$ref == '#/components/schemas/RequestType2_4_0'
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiInheritedPojoControllerSpec.groovy:1598]: unwrapped allOf

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy [hunk 7]
```diff
@@ -183,7 +188,8 @@ class MyBean {}
         requestBody.content['multipart/form-data'].schema.properties['files']
         requestBody.content['multipart/form-data'].schema.properties['files'] instanceof ArraySchema
         requestBody.content['multipart/form-data'].schema.properties['files'].description == 'The streaming files.'
-        requestBody.content['multipart/form-data'].schema.properties['files'].items instanceof BinarySchema
+        requestBody.content['multipart/form-data'].schema.properties['files'].items.type == 'string'
+        requestBody.content['multipart/form-data'].schema.properties['files'].items.format == 'binary'
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy:192]: now we have Schema instance, need to check type and format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy [hunk 8]
```diff
@@ -160,10 +163,12 @@ class MyBean {}
...
         requestBody.content['multipart/form-data'].schema.type == 'object'
         requestBody.content['multipart/form-data'].schema.properties['file1']
-        requestBody.content['multipart/form-data'].schema.properties['file1'] instanceof BinarySchema
+        requestBody.content['multipart/form-data'].schema.properties['file1'].type == 'string'
+        requestBody.content['multipart/form-data'].schema.properties['file1'].format == 'binary'
         requestBody.content['multipart/form-data'].schema.properties['file1'].description == "The streaming file 1."
         requestBody.content['multipart/form-data'].schema.properties['file2']
-        requestBody.content['multipart/form-data'].schema.properties['file2'] instanceof BinarySchema
+        requestBody.content['multipart/form-data'].schema.properties['file2'].type == 'string'
+        requestBody.content['multipart/form-data'].schema.properties['file2'].format == 'binary'
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy:171]: now we have Schema instance, need to check type and format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy [hunk 9]
```diff
@@ -160,10 +163,12 @@ class MyBean {}
         requestBody.content['multipart/form-data'].schema
         requestBody.content['multipart/form-data'].schema.type == 'object'
         requestBody.content['multipart/form-data'].schema.properties['file1']
-        requestBody.content['multipart/form-data'].schema.properties['file1'] instanceof BinarySchema
+        requestBody.content['multipart/form-data'].schema.properties['file1'].type == 'string'
+        requestBody.content['multipart/form-data'].schema.properties['file1'].format == 'binary'
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy:167]: now we have Schema instance, need to check type and format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy [hunk 10]
```diff
@@ -142,7 +144,8 @@ class MyBean {}
         requestBody.content['multipart/form-data'].schema
         requestBody.content['multipart/form-data'].schema.type == 'object'
         requestBody.content['multipart/form-data'].schema.properties['file']
-        requestBody.content['multipart/form-data'].schema.properties['file'] instanceof BinarySchema
+        requestBody.content['multipart/form-data'].schema.properties['file'].type == 'string'
+        requestBody.content['multipart/form-data'].schema.properties['file'].format == 'binary'
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy:148]: now we have Schema instance, need to check type and format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy [hunk 11]
```diff
@@ -124,7 +125,8 @@ class MyBean {}
...
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy:129]: now we have Schema instance, need to check type and format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy [hunk 12]
```diff
@@ -106,7 +106,8 @@ class MyBean {}
...
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadSpec.groovy:110]: now we have Schema instance, need to check type and format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadBodyParameterSpec.groovy [hunk 13]
```diff
@@ -274,7 +279,8 @@ class MyBean {}
         requestBody.content.size() == 1
         requestBody.content['multipart/form-data'].schema
         requestBody.content['multipart/form-data'].schema instanceof ArraySchema
-        requestBody.content['multipart/form-data'].schema.items instanceof BinarySchema
+        requestBody.content['multipart/form-data'].schema.items.type == 'string'
+        requestBody.content['multipart/form-data'].schema.items.format == 'binary'
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadBodyParameterSpec.groovy:283]: now we have Schema instance, need to check type and format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadBodyParameterSpec.groovy [hunk 14]
```diff
@@ -150,10 +153,12 @@ class MyBean {}
         requestBody.content['multipart/form-data'].schema
         requestBody.content['multipart/form-data'].schema.type == 'object'
         requestBody.content['multipart/form-data'].schema.properties['file1']
-        requestBody.content['multipart/form-data'].schema.properties['file1'] instanceof BinarySchema
+        requestBody.content['multipart/form-data'].schema.properties['file1'].type == 'string'
+        requestBody.content['multipart/form-data'].schema.properties['file1'].format == 'binary'
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadBodyParameterSpec.groovy:157]: now we have Schema instance, need to check type and format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadBodyParameterSpec.groovy [hunk 15]
```diff
@@ -150,10 +153,12 @@ class MyBean {}
...
         requestBody.content['multipart/form-data'].schema.properties['file1'].description == "Streaming File 1."
         requestBody.content['multipart/form-data'].schema.properties['file2']
-        requestBody.content['multipart/form-data'].schema.properties['file2'] instanceof BinarySchema
+        requestBody.content['multipart/form-data'].schema.properties['file2'].type == 'string'
+        requestBody.content['multipart/form-data'].schema.properties['file2'].format == 'binary'
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadBodyParameterSpec.groovy:161]: now we have Schema instance, need to check type and format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadBodyParameterSpec.groovy [hunk 16]
```diff
@@ -114,7 +115,8 @@ class MyBean {}
         requestBody.content['multipart/form-data'].schema
         requestBody.content['multipart/form-data'].schema.type == 'object'
         requestBody.content['multipart/form-data'].schema.properties['file']
-        requestBody.content['multipart/form-data'].schema.properties['file'] instanceof BinarySchema
+        requestBody.content['multipart/form-data'].schema.properties['file'].type == 'string'
+        requestBody.content['multipart/form-data'].schema.properties['file'].format == 'binary'
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadBodyParameterSpec.groovy:119]: now we have Schema instance, need to check type and format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadBodyParameterSpec.groovy [hunk 17]
```diff
@@ -96,7 +96,8 @@ class MyBean {}
...
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiFileUploadBodyParameterSpec.groovy:100]: now we have Schema instance, need to check type and format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiBasicSchemaSpec.groovy [hunk 18]
```diff
@@ -1428,13 +1430,14 @@ public class MyBean {}
...
+        schema.properties.date.default == OffsetDateTime.parse('2007-12-03T10:15:30+01:00')
         schema.properties.date.type == 'string'
         schema.properties.date.format == 'date-time'
 
-        schema.properties.mySubObject.default == 'myDefault3'
-        schema.properties.mySubObject.type == null
-        schema.properties.mySubObject.format == null
+        schema.properties.mySubObject.allOf.get(1).default == 'myDefault3'
+        schema.properties.mySubObject.allOf.get(1).type == null
+        schema.properties.mySubObject.allOf.get(1).format == null
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiBasicSchemaSpec.groovy:1440]: Now this field in allOf block

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiBasicSchemaSpec.groovy [hunk 19]
```diff
@@ -1321,7 +1323,7 @@ class DemoData {
     private URL url;
     @Schema(defaultValue = "274191c9-c176-4b1c-8263-1b658cbdc7fc")
     private UUID uuid;
-    @Schema(defaultValue = "Jan 12, 1952")
+    @Schema(defaultValue = "2007-12-03T10:15:30+01:00")
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiBasicSchemaSpec.groovy:1326]: swagger support only offset date time format

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiArraySchemaSpec.groovy [hunk 20]
```diff
@@ -72,49 +72,49 @@ class MyBean {}
...
-        openAPI.components.schemas['Pets'].properties['nestedIdArray'].description == 'a nested array of Ids'
-        openAPI.components.schemas['Pets'].properties['nestedIdArray'].items.items.format == 'int64'
-
-        openAPI.components.schemas['Pets'].properties['idArrayList'].description == 'a list of nested Ids'
-        openAPI.components.schemas['Pets'].properties['idArrayList'].items.items.format == 'int64'
-
-        openAPI.components.schemas['Pets'].properties['idListArray'].description == 'an array of nested Ids'
-        openAPI.components.schemas['Pets'].properties['idListArray'].items.items.format == 'int64'
+        petSchema.description == 'Pets'
+        petSchema.properties['pets'].nullable == false
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiArraySchemaSpec.groovy:81]: petSchema = openAPI.components.schemas['Pets']

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiApplicationVisitorSpec.groovy [hunk 21]
```diff
@@ -918,7 +899,6 @@ class MyBean {}
 ''')
         then: "the state is correct"
         !Utils.testReference
-        !Utils.testReferenceAfterPlaceholders
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiApplicationVisitorSpec.groovy:921]: Utils.testReferenceAfterPlaceholders - this filed deleted and we need to check testReference always

File: openapi/src/test/groovy/io/micronaut/openapi/MyJaxbElement4.java [hunk 22]
```diff
@@ -0,0 +1,90 @@
+package io.micronaut.openapi;
+
+import javax.validation.constraints.NotNull;
+import javax.validation.constraints.Pattern;
+import javax.validation.constraints.Size;
+
+import io.swagger.v3.oas.annotations.media.Schema;
+
+public class MyJaxbElement4 {
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/MyJaxbElement4.java:9]: `@Schema(oneOf = {DiscountSizeOpenApi.class, DiscountFixedOpenApi.class, MultiplierSizeOpenApi.class})` 
Test for this block. Fixed processing in this PR

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiRecordsSpec.groovy [hunk 23]
```diff
@@ -84,7 +83,6 @@ class MyBean {}
         then:
         openAPI.components.schemas
         openAPI.components.schemas.size() == 1
-        openAPI.components.schemas['Person'].name == 'Person'
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiRecordsSpec.groovy:87]: schema name always null. Schema name contains in schemas map

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiSchemaInContentSpec.groovy [hunk 24]
```diff
@@ -36,7 +36,7 @@ class OpenApiController {
             description = "This is description",
             tags = "Normalize",
             responses = {
-                    @ApiResponse(responseCode = "200", description = "Desc1", content = @Content(mediaType = MediaType.ALL, schema = @Schema(type = "blob", format = "binary"))),
+                    @ApiResponse(responseCode = "200", description = "Desc1", content = @Content(mediaType = MediaType.ALL, schema = @Schema(type = "string", format = "binary"))),
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiSchemaInContentSpec.groovy:39]: we can't use custom values for type field

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenapiCustomSchemaSpec.groovy [hunk 25]
```diff
@@ -35,6 +35,7 @@ class MyDto {
     private JAXBElement<? extends XmlElement> xmlElement;
     private JAXBElement<? extends XmlElement2> xmlElement2;
     public JAXBElement<? extends XmlElement3> xmlElement3;
+    public JAXBElement<? extends XmlElement4> xmlElement4;
```
Comment (altro3) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenapiCustomSchemaSpec.groovy:38]: new test case for oneOf block

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiParameterSchemaSpec.groovy (hunk 3 above)
Comment (dstepanov) [reply openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiParameterSchemaSpec.groovy:93]: We support no-annotation parameters, which can also be query value, but for proper OpenApi it might be needed

Comment (altro3) [reply openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiParameterSchemaSpec.groovy:93]: In this test before my hanges in field `in` was null value

**Review by dstepanov — APPROVED** [review]

File: openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiApplicationVisitorSpec.groovy [hunk 26]
```diff
@@ -775,24 +770,12 @@ class MyBean {}
...
-
-        def withRef = openAPI.components.securitySchemes['schemeWithRef']
-        withRef.type == null
-        withRef.in == null
-        withRef.name == null
-        withRef.description == null
-        withRef.openIdConnectUrl == null
-        withRef.bearerFormat == null
-        withRef.flows == null
-        withRef.scheme == null
```
Comment (graemerocher) [inline openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiApplicationVisitorSpec.groovy:787]: Why was this deleted?

Comment (altro3) [reply openapi/src/test/groovy/io/micronaut/openapi/visitor/OpenApiApplicationVisitorSpec.groovy:787]: @graemerocher Because the swagger doesn't really support links to other security schemes. Therefore, when you try to deserialize a swagger file with such a scheme, it will be ignored, because. type is not specified. See class SecuritySchemeDeserializer:
изображение
That is, this test case is not correct.
